Modules/gcmodule.c

```diff
@@ -1111,17 +1128,21 @@ collect(struct _gc_runtime_state *state, int generation,
     /* Call tp_finalize on objects which have one. */
     finalize_garbage(&unreachable);
 
-    if (check_garbage(&unreachable)) { // clear PREV_MASK_COLLECTING here
-        gc_list_merge(&unreachable, old);
-    }
-    else {
-        /* Call tp_clear on objects in the unreachable set.  This will cause
-         * the reference cycles to be broken.  It may also cause some objects
-         * in finalizers to be freed.
-         */
-        m += gc_list_size(&unreachable);
-        delete_garbage(state, &unreachable, old);
-    }
+    // The deduce_unreachable function will move all objects that are still
+    // reachable from the outside into the still_unreachable set.
+    PyGC_Head still_unreachable;
+    gc_list_init(&still_unreachable);
+    if (deduce_unreachable(&unreachable, &still_unreachable)) { // clear PREV_MASK_COLLECTING here
+        // Move every resurrected object (and the ones reachable from them) into
+        // the yoing list for future collection.
+        gc_list_merge(&still_unreachable, young);
```

Comment:
`still_unreachable` throughout the last several lines should be, e.g., `now_reachable` instead, yes?